tests/test_models.py

```diff
@@ -265,3 +268,17 @@ def test_dispatch_model_solve_case4(technology_set_4, net_demand):
                              model.results.Battery_charge)
     assert (total_gen - net_demand.sum()) == pytest.approx(0, abs=TOL)
     assert binary_charging == pytest.approx(0, abs=TOL)
+
+def test_dispatch_model_solve_case5(technology_set_4, net_demand):
+    """
+    Tests that the curtailment technology behaves as expected.
+    """
```

Comment:
You have 4 tests for not using curtailment. Do you think only 1 with curtailment is sufficient?

Reply:
Curtailment just removes excess electricity from the grid (e.g. shooting it into the ground), it's energy that gets wasted. Typically used in a high renewable context, like when net demand goes negative. If there isn't enough storage capacity, it has to be curtailed. I just test it by saying here's a nuclear reactor that operates with constant power and rather than ramp up or down, just curtail it.

The "test" is whether or not curtailment is used as intended. If curtailment is enabled the total supply of electricity is

supply = sum(generation from all technologies) - curtailed electricity

and the constraint is that

supply = demand +/- margin.

If it's not enabled, then the supply is just the sum of generation over all technologies. In this case, there are four "negative" tests, which check that no energy is accidentally being curtailed, and the one positive test makes sure that excess energy is removed.

Reply:
And this test is using just a single technology? What about if you have multiple technologies? Can you have one technology that uses curtailment and one that doesn't?

Reply:
The *model* uses curtailment. The way I implemented it, curtailment *is* a technology, like nuclear or solar or natural gas. So the model chooses how much to curtail the same way it chooses how much nuclear to produce. It's not something that gets used by a single technology.

Reply:
Ah. I did not pick up on that distinction. That makes sense. I will still encourage you to think about if any new tests need to be added in for edge or corner cases. Instinct tells me that 1 test isn't enough, but it may actually be.